shavee-bin/src/main.rs

```diff
@@ -1,90 +1,54 @@
+    exit(match run(args){
+        Ok(()) => 0,
         Err(error) => {
-            eprintln!("Error: Failed to read Password");
             eprintln!("Error: {}", error);
-            exit(1)
+            1
         }
-    };
+    });
+}
+
+
+fn run(args: Sargs) -> Result<(), Box<dyn Error>> {
+
+    let pass = rpassword::prompt_password_stderr("Dataset Password: ")
+        .map_err(|e| e.to_string())?;
 
-    match args.umode.as_str() {
-        "yubikey" => match args.mode.as_str() {
-            "print" => match print_mode_yubi(pass, args.yslot) {
-                Ok(()) => exit(0),
-                Err(e) => {
-                    eprintln!("Error: {}", e);
-                    exit(1)
-                }
-            },
-            "pam" | "mount" => match unlock_zfs_yubi(pass, args.dataset, args.yslot) {
-                Ok(()) => exit(0),
-                Err(e) => {
-                    eprintln!("Error: {}", e);
-                    exit(1)
-                }
-            },
-            "create" => match create_zfs_yubi(pass, args.dataset, args.yslot) {
-                Ok(()) => exit(0),
-                Err(e) => {
-                    eprintln!("Error: {}", e);
-                    exit(1)
-                }
-            },
-            _ => unreachable!(),
+    Ok(match args.umode {
+        Umode::Yubikey => match args.mode {
+            Mode::Print => print_mode_yubi(pass, args.yslot)?,
+            Mode::Mount => unlock_zfs_yubi(pass, args.dataset, args.yslot)?,
+            Mode::Create => create_zfs_yubi(pass, args.dataset, args.yslot)?,
         },
-        "file" => match args.mode.as_str() {
-            "print" => match print_mode_file(pass, &args.file, args.port) {
-                Ok(()) => exit(0),
-                Err(e) => {
-                    eprintln!("Error: {}", e);
-                    exit(1)
-                }
-            },
-            "pam" | "mount" => match unlock_zfs_file(pass, args.file, args.dataset, args.port) {
-                Ok(()) => exit(0),
-                Err(e) => {
-                    eprintln!("Error: {}", e);
-                    exit(1)
-                }
-            },
-            "create" => match create_zfs_file(pass, args.file, args.dataset, args.port) {
-                Ok(()) => exit(0),
-                Err(e) => {
-                    eprintln!("Error: {}", e);
-                    exit(1)
-                }
-            },
-            _ => unreachable!(),
+        Umode::File => match args.mode {
+            Mode::Print => print_mode_file(pass, args.file, args.port)?,
+            Mode::Mount => unlock_zfs_file(pass, args.file, args.dataset, args.port)?,
+            Mode::Create => create_zfs_file(pass, args.file, args.dataset, args.port)?,
         },
-        "password" => {
-            let key = hash_argon2(pass.into_bytes()).unwrap();
+        Umode::Password => {
+            let key =  hash_argon2(pass.into_bytes())?;
```

Comment:
This is a repeat of `file_key_calculation()` and instead it is possible to call `file_key_calculation()` this way:

```Rust
let key =  file_key_calculation(pass, None, None)?;
```

However if the above is implemented, `file_key_calculation()` must be changed to
* If `file` is `None` then doesn't call `get_filehash()` nor concatenation trait.